Alignment/include/ActsAlignment/Kernel/detail/AlignmentEngine.hpp

```diff
@@ -0,0 +1,302 @@
+
+  // The alignable surfaces on the track and their indices in both the global
+  // alignable surfaces pool and those relevant with this track
+  std::unordered_map<const Surface*, std::pair<size_t, size_t>> alignedSurfaces;
+};
+
+/// Reset some columns of the alignment to bound derivative to zero if the
+/// relevant degree of freedom is fixed
+///
+/// @param alignToBound The alignment to bound parameters derivative
+/// @param mask The alignment mask
+void resetAlignmentDerivative(Acts::AlignmentToBoundMatrix& alignToBound,
+                              AlignmentMask mask);
+
+///
+/// Calculate the first and second derivative of chi2 w.r.t. alignment
+/// parameters for a single track
+///
+/// Suppose there are n measurements on the track, and m (m<=n) of them are on
+/// alignable surface, then eAlignmentSize*m alignment parameters
+/// will be involved for this particular track, i.e. this track will contribute
+/// to at most eAlignmentSize*m*2 elements of the full chi2
+/// second derivative matrix
+///
+/// @tparam source_link_t The source link type of the trajectory
+/// @tparam parameters_t The track parameters type
+///
+/// @param gctx The current geometry context object
+/// @param multiTraj The MultiTrajectory containing the trajectory to be
+/// investigated
+/// @param entryIndex The trajectory entry index
+/// @param globalTrackParamsCov The global track parameters covariance for a
+/// single track and the starting row/column for smoothed states. This contains
+/// all smoothed track states including those non-measurement states. Selection
+/// of certain rows/columns for measurement states is needed.
+/// @param idxedAlignSurfaces The indexed surfaces to be aligned
+///
+/// @return The track alignment state containing fundamental alignment
+/// ingredients
+template <typename source_link_t, typename parameters_t = BoundTrackParameters>
+TrackAlignmentState trackAlignmentState(
+    const GeometryContext& gctx,
+    const Acts::MultiTrajectory<source_link_t>& multiTraj,
+    const size_t& entryIndex,
+    const std::pair<ActsDynamicMatrix, std::unordered_map<size_t, size_t>>&
+        globalTrackParamsCov,
+    const std::unordered_map<const Surface*, size_t>& idxedAlignSurfaces,
+    const AlignmentMask& alignMask) {
+  using CovMatrix = typename parameters_t::CovarianceMatrix;
+
+  // Construct an alignment state
+  TrackAlignmentState alignState;
+
+  // Remember the index within the trajectory and whether it's alignable
+  std::vector<std::pair<size_t, bool>> measurementStates;
+  measurementStates.reserve(15);
+  // Number of smoothed states on the track
+  size_t nSmoothedStates = 0;
+  // Number of alignable surfaces on the track
+  size_t nAlignSurfaces = 0;
+
+  // Visit the track states on the track
+  multiTraj.visitBackwards(entryIndex, [&](const auto& ts) {
+    // Remember the number of smoothed states
+    if (ts.hasSmoothed()) {
+      nSmoothedStates++;
+    } else {
+      // @FIXME: is this possible?
```

Comment:
Good question: if the smoothing generated an outlier, what would you do then?

Reply:
It depends on what we mean by an outlier during smoothing. We don't have the logic to generate an outlier in smoothing yet, so this `else` should in principle never happen.
Now suppose we have the possibility to get an outlier during the smoothing in the future, I expect we still have the smoothing parameters on this state? If no, we could just don't use the measurement on this state. 
So I guess we don't have to worry about this too much at the moment...